pkg/kubelet/kuberuntime/kuberuntime_manager.go

```diff
@@ -543,6 +547,272 @@ func containerSucceeded(c *v1.Container, podStatus *kubecontainer.PodStatus) boo
+		// TODO(vinaykul): Store runtime.Container.Resources in kubeContainerStatus
+		//    instead of v1.Resources, translate for below update operations
+		if cInfo.kubeContainerStatus.Resources.Limits != nil {
+			currentLimits = cInfo.kubeContainerStatus.Resources.Limits.DeepCopy()
+		}
+		if cInfo.kubeContainerStatus.Resources.Requests != nil {
+			currentRequests = cInfo.kubeContainerStatus.Resources.Requests.DeepCopy()
+		}
+		switch resourceKind {
+		case ResizeCPULimit:
+			container.Resources.Limits = v1.ResourceList{
+				v1.ResourceCPU:    *desiredLimits.Cpu(),
+				v1.ResourceMemory: *currentLimits.Memory(),
+			}
+			containerStatus.ResourcesAllocated = v1.ResourceList{
+				v1.ResourceCPU:    *currentRequests.Cpu(),
+				v1.ResourceMemory: *currentRequests.Memory(),
+			}
+		case ResizeCPURequest:
+			container.Resources.Limits = v1.ResourceList{
+				v1.ResourceCPU:    *currentLimits.Cpu(),
+				v1.ResourceMemory: *currentLimits.Memory(),
+			}
+			containerStatus.ResourcesAllocated = v1.ResourceList{
+				v1.ResourceCPU:    *desiredRequests.Cpu(),
+				v1.ResourceMemory: *currentRequests.Memory(),
+			}
+		case ResizeMemoryLimit:
+			container.Resources.Limits = v1.ResourceList{
+				v1.ResourceCPU:    *currentLimits.Cpu(),
+				v1.ResourceMemory: *desiredLimits.Memory(),
+			}
+			containerStatus.ResourcesAllocated = v1.ResourceList{
+				v1.ResourceCPU:    *currentRequests.Cpu(),
+				v1.ResourceMemory: *currentRequests.Memory(),
+			}
+		}
+		if err := m.updateContainerResources(pod, container, cInfo.kubeContainerStatus.ID); err != nil {
+			// Log error and abort as container updates need to succeed in the order determined by computePodResizeAction
+			// The recovery path is for SyncPod to keep retrying at later times until it succeeds.
+			klog.ErrorS(err, "updateContainerResources failed", "container", container.Name, "cID", cInfo.kubeContainerStatus.ID,
+				"pod", format.Pod(pod), "resourceKind", resourceKind)
+			return err
+		}
+		updatedContainersMap[cInfo.kubeContainerStatus.ID] = cInfo.kubeContainerStatus
+	}
+
+	if len(updatedContainersMap) > 0 {
+		// Update pod cache runtime container status (cInfo.kubeContainerStatus) with latest values from ContainerStatus API.
+		newPodStatus, err := m.GetPodStatus(podStatus.ID, pod.Name, pod.Namespace)
+		if err != nil {
+			klog.ErrorS(err, "GetPodStatus failed", "pod", format.Pod(pod))
+			return err
+		}
+		for _, newContainerStatus := range newPodStatus.ContainerStatuses {
+			if c, found := updatedContainersMap[newContainerStatus.ID]; found {
+				if newContainerStatus.Resources.Limits == nil && newContainerStatus.Resources.Requests == nil {
+					klog.ErrorS(err, "runtime failed to report resources", "pod", format.Pod(pod))
+					return fmt.Errorf("runtime failed to report resources for pod %s", pod.Name)
+				}
+				if newContainerStatus.Resources.Limits != nil {
+					c.Resources.Limits = newContainerStatus.Resources.Limits.DeepCopy()
```

Comment:
I don't quite understand, why do you need to update the status in `containersToUpdate`?

Reply:
containersToUpdate holds the reference to the kubeContainerStatus that is associated with the container (queried by podStatus.FindContainerStatusByName() function in computePodActions) that was just resized. 

So when ContainerStatus CRI API (called via GetPodStatus) returns the latest config after UpdateContainerResources was called, we refresh the values in kubeContainerStatus.

Later, this gets picked up by generateAPIstatus in kubelet_pods.

Reply:
> Later, this gets picked up by generateAPIstatus in kubelet_pods.

I'm not sure how this would actually work. 

`generateAPIstatus` is called before `SyncPod`. And the pod status it uses should come from the pod cache.

How would your update here affect that?

Reply:
It is a loop. 
- Resize decision and allocation get posted after HandlePodUpdate accepts the resize so the user quickly knows what's the status.
- CRI update happens in SyncLoop, and ContainerStatus response is gathered and updated on the kubeContainerStatus pointer.
- In the next iteration of the loop that invokes convertToApiPodStatus, apiPodStatus is updated.

Here's a trace of how this plays out:
```vim
root@fw0000359:~/go/src/k8s.io/kubernetes-rfpvs/pkg/kubelet# cat /tmp/kubelet.log |grep VDBG
E0707 16:10:02.419155    8083 kubelet_pods.go:1646] <<<<<-----VDBG_CONVERT_TO_APIPODSTATUS: POD 2pod PODSTATUS.RESIZE: ''
E0707 16:10:03.604696    8083 kubelet_pods.go:1646] <<<<<-----VDBG_CONVERT_TO_APIPODSTATUS: POD 2pod PODSTATUS.RESIZE: ''
E0707 16:10:05.618958    8083 kubelet_pods.go:1924] <<<<<-----VDBG_CONVERT_TO_APICONTAINERSTATUS: POD 2pod CTR: stress KUBE_CONTAINERSTATUS_RESOURCES: {map[cpu:{{400 -3} {<nil>}  DecimalSI} memory:{{419430400 0} {<nil>}  BinarySI}] map[cpu:{{200 -3} {<nil>}  DecimalSI}]} STATUS.RESOURCES: &ResourceRequirements{Limits:ResourceList{cpu: {{400 -3} {<nil>}  DecimalSI},memory: {{419430400 0} {<nil>}  BinarySI},},Requests:ResourceList{cpu: {{200 -3} {<nil>}  DecimalSI},memory: {{209715200 0} {<nil>}  BinarySI},},}
E0707 16:10:05.622164    8083 kubelet_pods.go:1646] <<<<<-----VDBG_CONVERT_TO_APIPODSTATUS: POD 2pod PODSTATUS.RESIZE: ''
E0707 16:10:06.623189    8083 kubelet_pods.go:1924] <<<<<-----VDBG_CONVERT_TO_APICONTAINERSTATUS: POD 2pod CTR: stress KUBE_CONTAINERSTATUS_RESOURCES: {map[cpu:{{400 -3} {<nil>}  DecimalSI} memory:{{419430400 0} {<nil>}  BinarySI}] map[cpu:{{200 -3} {<nil>}  DecimalSI}]} STATUS.RESOURCES: &ResourceRequirements{Limits:ResourceList{cpu: {{400 -3} {<nil>}  DecimalSI},memory: {{419430400 0} {<nil>}  BinarySI},},Requests:ResourceList{cpu: {{200 -3} {<nil>}  DecimalSI},memory: {{209715200 0} {<nil>}  BinarySI},},}
E0707 16:10:06.626254    8083 kubelet_pods.go:1646] <<<<<-----VDBG_CONVERT_TO_APIPODSTATUS: POD 2pod PODSTATUS.RESIZE: ''
E0707 16:10:53.399070    8083 kubelet_pods.go:1924] <<<<<-----VDBG_CONVERT_TO_APICONTAINERSTATUS: POD 2pod CTR: stress KUBE_CONTAINERSTATUS_RESOURCES: {map[cpu:{{400 -3} {<nil>}  DecimalSI} memory:{{419430400 0} {<nil>}  BinarySI}] map[cpu:{{200 -3} {<nil>}  DecimalSI}]} STATUS.RESOURCES: &ResourceRequirements{Limits:ResourceList{cpu: {{400 -3} {<nil>}  DecimalSI},memory: {{419430400 0} {<nil>}  BinarySI},},Requests:ResourceList{cpu: {{200 -3} {<nil>}  DecimalSI},memory: {{209715200 0} {<nil>}  BinarySI},},}
E0707 16:10:53.399586    8083 kubelet_pods.go:1646] <<<<<-----VDBG_CONVERT_TO_APIPODSTATUS: POD 2pod PODSTATUS.RESIZE: ''
E0707 16:10:53.400165    8083 kubelet.go:2334] >>>>>-----VDBG_HANDLE_POD_RESOURCE_RESIZE: POD 2pod POD_RESIZE: 'Proposed'
E0707 16:10:53.402773    8083 kubelet.go:2354] <<<<<-----VDBG_HANDLE_POD_RESOURCE_RESIZE: POD 2pod POD_RESIZE: 'InProgress'
E0707 16:10:53.449569    8083 kuberuntime_manager.go:788] >>>>>-----VDBG_UPDATE_POD_CONTAINER_RESOURCES: POD 2pod CTR: stress CTR_RESOURCES: {map[cpu:{{400 -3} {<nil>}  DecimalSI} memory:{{419430400 0} {<nil>}  BinarySI}] map[cpu:{{300 -3} {<nil>} 300m DecimalSI} memory:{{209715200 0} {<nil>}  BinarySI}]} KUBE_CONTAINER_RESOURCES: {map[cpu:{{400 -3} {<nil>}  DecimalSI} memory:{{419430400 0} {<nil>}  BinarySI}] map[cpu:{{200 -3} {<nil>}  DecimalSI}]}
E0707 16:10:53.456889    8083 kuberuntime_manager.go:811] <<<<<-----VDBG_UPDATE_POD_CONTAINER_RESOURCES: POD 2pod CTR: stress KUBE_CONTAINER_RESOURCES: {map[cpu:{{400 -3} {<nil>}  DecimalSI} memory:{{419430400 0} {<nil>}  BinarySI}] map[cpu:{{300 -3} {<nil>}  DecimalSI}]}
E0707 16:12:02.584007    8083 kubelet_pods.go:1924] <<<<<-----VDBG_CONVERT_TO_APICONTAINERSTATUS: POD 2pod CTR: stress KUBE_CONTAINERSTATUS_RESOURCES: {map[cpu:{{400 -3} {<nil>}  DecimalSI} memory:{{419430400 0} {<nil>}  BinarySI}] map[cpu:{{300 -3} {<nil>}  DecimalSI}]} STATUS.RESOURCES: &ResourceRequirements{Limits:ResourceList{cpu: {{400 -3} {<nil>}  DecimalSI},memory: {{419430400 0} {<nil>}  BinarySI},},Requests:ResourceList{cpu: {{300 -3} {<nil>}  DecimalSI},memory: {{209715200 0} {<nil>}  BinarySI},},}
E0707 16:12:02.587222    8083 kubelet_pods.go:1646] <<<<<-----VDBG_CONVERT_TO_APIPODSTATUS: POD 2pod PODSTATUS.RESIZE: ''
root@fw0000359:~/go/src/k8s.io/kubernetes-rfpvs/pkg/kubelet#
```

Reply:
> In the next iteration of the loop that invokes convertToApiPodStatus, apiPodStatus is updated.

I don't think it is something we should depend on.

The pod status comes from the pod cache https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/pod_workers.go#L167

I don't think we are supposed to update it here, and it is not correct either, your edition will be overwritten in the next pod cache update. The pod cache is expected to reflect the status from the container runtime.

Reply:
I'll probably need your help with this. There are deterministic ways to arrive at ContainerStateRunning / ContainerStateExited etc but you can't really have a ContainerStateUpdated as a current state. Everything I'm thinking of seems like a hack. My brain is a little fried now so .. :|

Reply:
TODO added to code. I'll re-test a small fix and update tonight.

Reply:
> Everything I'm thinking of seems like a hack.

If we are OK with a 10 second delay, we can rely on the periodical pod worker sync to update the status in the next iteration.

In beta, we can think about how to better propagate this. We can requeue an update in the pod worker, or change PLEG to be based on container runtime event (update has an event).

Reply:
I don't think a TODO is enough here. :/ Can we remove the code, and rely on the 10 second periodical sync to update the status? Will it cause any problem?

Reply:
I remember why I did this now... when implementing the previous approved version of this KEP, I discovered that one of the conditions we use to determine ContainersToUpdate in computePodActions is container.Resources != kubeContainerStatus.Resources . Doing this prevents spurious UpdateContainerResources calls.

Would 10s worker sync be good enough to avoid unnecessary UpdateContainerResources? .. It may be.. I'll have to experiment with it and see, I'll try it out today evening.

Reply:
> Would 10s worker sync be good enough to avoid unnecessary UpdateContainerResources?

The pod worker makes sure that the pod status in the next syncloop must be newer than the previous pod cache, so in most cases the `kubeContainerStatus` should have been updated (especially after 10 seconds). Even if it is not updated yet, it should be fine to run another update, it is idempotent anyway.

Reply:
I tried removing the above code that updates kubeContainerStatus. Resource updates are not captured. Cursorily reading PLEG code it is likely because it does not find an event that gets it to call GetPodStatus to refresh kubeContainerStatus. 

So if I just remove this code without adding some mechanism elsewhere, we are in endless UpdateContainerResources loop.

Reply:
> Cursorily reading PLEG code it is likely because it does not find an event that gets it to call GetPodStatus to refresh kubeContainerStatus.

That is a missing piece in the current design.

To work around this for now, we can call a `GetPodStatus` before the diff to get the up-to-date status. And better address this in beta.

Reply:
Yes, GetPodStatus before diff is a better idea as it accounts for potential lag in CRI actuating the resize.

Reply:
@Random-Liu I tried moving GetPodStatus call to before diff, but that move breaks things e.g. If I update CPU and memory of a guaranteed pod, the update algorithm breaks down the requested update into discrete cpu and memory updates to handle the cases where they may be ordered differently.

Let's take the example where memory limit and cpu limits are updated by VPA. If mem update fails, we bail and retry later. If it is successful, then the next call to UpdateContainerResources to update the CPU needs to use the already updated values of memory in order to avoid bouncing the memory back to previous values. 

So we need to call GetPodStatus after UpdateContainerResources in order to get the latest.

I found a way to easily update PLEG to query GetPodStatus (we add a new bool field in kubePodStatus that indicates Resize is in progress. When PLEG sees this bit set in relist, it will call updateCache. This bit is cleared when Spec.Resources == Status.Resources. I'm not sure if this modification is necessary anymore since we need to keep the current code.

I'll rebase the PR and send it for review. We can take another look into it during this coming week to see if anything can be improved.

Reply:
> Let's take the example where memory limit and cpu limits are updated by VPA. If mem update fails, we bail and retry later. If it is successful, then the next call to UpdateContainerResources to update the CPU needs to use the already updated values of memory in order to avoid bouncing the memory back to previous values.

I don't quite understand this.

Based on my understanding, the desired memory limit should come from the desired state, and should not change. Why would it bounce the memory back?

Calling `GetPodStatus` before diff just make sure that if memory has already been updated, no diff is generated, and we don't need to update it again.

However, even if `GetPodStatus` doesn't return the up-to-date value, the update runs, it should still try to update the memory to the desired value, right?

Why would it bounce it back?

Reply:
@Random-Liu I think we can call GetPodStatus before computePodResizeAction so that we are using latest kubeContainerStatus **in addition to** calling it after UpdateContainerResources as we currently do - I’ll make that change. Below is the case for why we need it after.. 

Scenario: Consider node 4 cpu, 4G mem running a single two-container guaranteed QoS pod P [C1[cpu=2, mem=2Gi], C2[cpu=2, men=2Gi]]
Let’s say VPA determines this pod needs resize as follows: [C1[cpu=3, mem=1G], C2[cpu=1, mem=3G]} and updates the PodSpec.

Node facts at this point:
- Pod resize is admissible as net resource needs are unchanged, so we “admit” the resize and set v1.ctrStatus resourcesAllocated and update checkpoint accordingly.
- We cannot update container as an atomic unit (UpdateContainerResources CRI API) .. if we updated C1 first, we are oversubscribed on CPU, if C2 first then we are oversubscribed on memory.
- We need to update by resource-type i.e memory, cpu and order it to apply decreases before increases. aka example update-list: C1-mem, C2-mem, C2-cpu, C1-cpu.
- v1.ContainerSpec=desired-resources , v1.ContainerStatus/runtime.kubeContainerStatus=current-resources.

This means in above example ordering, we call UpdateContainerResources as follows:
Call 1:  C1 [desired-mem=1G + current-cpu=2]
Call 2:  C2 [desired-mem=3G + current-cpu=2]
GetPodStatus()
Call 3:  C2 [current-mem=3G, desired-cpu=1]
Call 4:  C1 [current-mem=1G + current-cpu=3]

If we didn’t do GetPodStatus between call 2 and 3, we would be using stale current-mem=2G a.k.a toggle back. 

Now that you mention the case where GetPodStatus does not turn-around quickly, I think we should consider bringing my earlier implementation back - i.e if UpdateContainerResources didn’t fail, assume that desired-value for the resource-type was accepted by runtime, and use it in the next call i.e calls 3 & 4 above. 

If you recall, my earlier implementation handled the case where GetPodStatus didn’t return ContainerResources since this is a new field in CRI API so runtimes may lag in supporting it. (But it is a manageable problem given a limited number of runtime implementation folks we need to reach out to and push …  we can educate in the next KubeCon and require this desired behavior within a release or two before we exit beta).

Thoughts?

Reply:
In that case, why can't we invoke `GetPodStatus` before Call 3? In that case, we don't need to update the pod cache to preserve state, which is racy.

And calling it before `Call 3` is easier to understand as well, we need to do break glass call to get the up-to-date state before each resource re-allocation, which makes sense to me.

Reply:
> In that case, why can't we invoke GetPodStatus before Call 3 (or even before every resize call)? In that case, we don't need to update the pod cache to preserve state, which is racy.

That's what the current code does - invokes GetPodStatus after updating C1, C2 mem (and again after Call 4). See https://github.com/kubernetes/kubernetes/pull/102884/files#diff-0e1a8d22b0f1dc42d412749c6a916b4c76394ead739d4bfbf26564faedd74a2fR857

> And calling it before Call 3 is easier to understand as well: we need to use break glass call to get the up-to-date state before each resource re-allocation, which makes sense to me.

Do you mean calling GetPodStatus before **Call 1** to update kuberuntime ContainerStatus?


In any case, it works now because GetPodStatus turnaround is quick. However, we need to account for case where runtime is slow to actuate. 

I think to be safe, we need to go back to my earlier implementation i.e if UpdateContainerResources succeeds, desired-value was accepted, and treat it as "current-value" for subsequent UpdateContainerResources calls. Specifically, after Call 2 success, we assume C2.current-mem=3G for Call 3 (because GetPodStatus could still return 2G) .. thoughts?!

And another question: Is there a way to tell PLEG periodic relist to invoke updateCache (which calls GetPodStatus) only when resize is in progress? (i.e kuberuntime.ContainerStatus.Resources != v1.ContainerStatus.Resources)

Reply:
> However, we need to account for case where runtime is slow to actuate.

The PLEG is introduced for that. But now, because we can't use PLEG, I think it is fine to directly call `GetPodStatus` to make sure correctness.

> That's what the current code does - invokes GetPodStatus after updating C1, C2 mem (and again after Call 4).

I don't understand why you need to update `podStatus` here. That is updating a pointer from the pod cache, which is not thread safe.

If you need the up-to-date pod status, just call it before you need it, e.g. making resize decision, instead of trying to cache the status in `podStatus`.

Reply:
>  don't understand why you need to update podStatus here. That is updating a pointer from the pod cache, which is not thread safe.

I agree, and I want to solve this race now (before alpha commit). 

The issue is that there is a gap that needs to be bridged: v1.PodStatus / v1.ContainerStatuses.Resources is generated from runtime.ContainerStatuses.Resources which comes from podCache (correct me if I'm wrong about this). 

Two ways to update runtime.ContainerStatuses.Resoruces:
1. Update the pointer as I currently do (not thread-safe)
2. Make PLEG relist to call GetPodStatus and update cache (not efficient)

If we don't do one of the above two things, I discovered that v1.ContainerStatus.Resources is never updated (even though my local GetPodStatus before shows resources have been correctly updated and actuated). 

Is there a good way to get PLEG relist to call GetPodStatus only when resize is in progress (aka v1.Spec.Resources != v1.Status.Resources) ?

Reply:
Maybe as a stop-gap / temporary compromise, we can let relist call GetPodStatus every 60 seconds? 

So every 60s do this: 
```
@@ -226,6 +232,15 @@ func (g *GenericPLEG) relist() {
                                updateEvents(eventsByPodID, e)
                        }
                }
+
+              if g.cacheEnabled() && pod != nil {
+                               if err := g.updateCache(pod, pid); err != nil {
+                                       // Rely on updateCache calling GetPodStatus to log the actual error.
+                                       klog.V(4).ErrorS(err, "PLEG: Ignoring updateCache failure for pod being resized", "pod", klog.KRef(pod.Namespace, pod.Name))
+                               }
+               }
```

There's another problem where cgroup updates quickly but the actual v1.Status.Resources takes around 90s for it to be reflected - effectively rendering the observed resize very slow. I think that can be solved before beta along with finding a way to get PLEG to call GetPodStatus only when resize is in progress.

Reply:
@Random-Liu Please review commit https://github.com/kubernetes/kubernetes/commit/b76dc0e94ff9242f5b1f159fe564a4673d215523 , I redid the implementation in kuberuntime_manager to not update the pod-cache pointer, and added interface in PLEG that is called from kubelet syncPod when the runtime containers are being resized. It did simplify the KRT logic quite a bit. 

I need to update the unit test but I'll do that once we see review the new implementation and are good with the containersToUpdate struct and related changes, as UT depends on it.

Reply:
That looks better to me. As long as we make sure the logic only runs when pod is resized, it should be good enough for alpha.

Reply:
Great, thanks for the quick review. 

I think we have it it isolated.. I'll convert over the unit tests tonight and also double-check to verify all the operational code is contained inside the feature gate. Passing CI tests is hopefully a good sign.

Reply:
Unit tests fixed in latest commit.